pyrtl/simulation.py

```diff
@@ -793,15 +793,44 @@ def add_fast_step(self, fastsim):
         for wire_name in self.trace:
             self.trace[wire_name].append(fastsim.context[wire_name])
 
-    def print_trace(self, file=sys.stdout):
-        # TODO: make multi-digit values more readable and implement better alignment (issue #203)
+    def print_trace(self, file=sys.stdout, base=10, compact=False):
+        """
+        Prints a list of wires and their current values.
+        :param int base: the base the values are to be printed in
+        :param bool compact: whether to omit spaces in output lines
+        """
         if len(self.trace) == 0:
             raise PyrtlError('error, cannot print an empty trace')
-        maxlen = max(len(w) for w in self.trace)
-        for w in sorted(self.trace, key=_trace_sort_key):
-            file.write(' '.join([w.rjust(maxlen),
-                       ''.join(str(x) for x in self.trace[w])+'\n']))
-            file.flush()
+        if base not in (2, 8, 10, 16):
+            raise PyrtlError('please choose a valid base')
+
+        def dec(i):
+            return '0d' + str(i)
+
+        def octal(i):
+            return ('0o' + oct(i)[(0 if i == 0 else 1):] if sys.version_info.major == 2
+                    else oct(i))
+
+        bfunc = {2: bin, 8: octal, 10: dec, 16: hex}[base]
+        maxlenleft = max(len(w) for w in self.trace)
+        if compact:
+            for w in sorted(self.trace, key=_trace_sort_key):
+                file.write(' '.join([w.rjust(maxlenleft),
+                                     ''.join(bfunc(x)[2:] for x in self.trace[w]) + '\n']))
+                file.flush()
+
+        else:
+            maxlenval = max(len(bfunc(x)[2:])
+                            for w in self.trace for x in self.trace[w])
+            maxlenright = (maxlenval+1) * len(list(self.trace.values())[0])
+            file.write(' '*(maxlenleft-3) + "--- Values in base %d ---\n" % base)
+            for w in sorted(self.trace, key=_trace_sort_key):
+                file.write(w.ljust(maxlenleft)+'')
```

Comment:
Actually the python 2.x tests were throwing an error saying that `write()` expected a unicode and got a string instead, so I first used the `unicode()` method but that apparently is only implemented in python 2.x, so I found that adding `+''` converts it and takes care of the 2.x problem while still working in 3.x